spark-operator/src/main/java/org/apache/spark/k8s/operator/metrics/source/OperatorJosdkMetrics.java

```diff
@@ -0,0 +1,312 @@
+  private static final String SIZE = "size";
+
+  private final Clock clock;
+  private final MetricRegistry metricRegistry;
+
+  public OperatorJosdkMetrics() {
+    this.clock = new SystemClock();
+    this.metricRegistry = new MetricRegistry();
+  }
+
+  @Override
+  public String sourceName() {
+    return PREFIX;
+  }
+
+  @Override
+  public MetricRegistry metricRegistry() {
+    return metricRegistry;
+  }
+
+  @Override
+  public void controllerRegistered(Controller<? extends HasMetadata> controller) {
+    // no-op
+    log.debug("Controller has been registered");
+  }
+
+  @Override
+  public void receivedEvent(Event event, Map<String, Object> metadata) {
+    log.debug("received event {}, metadata {}", event, metadata);
+    if (event instanceof ResourceEvent) {
+      final ResourceAction action = ((ResourceEvent) event).getAction();
+      final Optional<Class<? extends BaseResource<?, ?, ?, ?, ?>>> resource =
+          getResourceClass(metadata);
+      final Optional<String> namespaceOptional = event.getRelatedCustomResourceID().getNamespace();
+      resource.ifPresent(
+          aClass -> getCounter(aClass, action.name().toLowerCase(), RESOURCE, EVENT).inc());
+      if (resource.isPresent() && namespaceOptional.isPresent()) {
+        getCounter(
+                resource.get(),
+                namespaceOptional.get(),
+                action.name().toLowerCase(),
+                RESOURCE,
+                EVENT)
+            .inc();
+      }
+    }
+  }
+
+  @Override
+  public <T> T timeControllerExecution(ControllerExecution<T> execution) throws Exception {
+    log.debug("Time controller execution");
+    final String name = execution.controllerName();
+    final ResourceID resourceID = execution.resourceID();
+    final Optional<String> namespaceOptional = resourceID.getNamespace();
+    final Map<String, Object> metadata = execution.metadata();
+    final Optional<Class<? extends BaseResource<?, ?, ?, ?, ?>>> resourceClass =
+        getResourceClass(metadata);
+    final String execName = execution.name();
+
+    long startTime = clock.getTimeMillis();
+    try {
+      T result = execution.execute();
+      final String successType = execution.successTypeName(result);
+      if (resourceClass.isPresent()) {
+        getHistogram(resourceClass.get(), name, execName, successType).update(toSeconds(startTime));
+        getCounter(resourceClass.get(), name, execName, SUCCESS, successType).inc();
+        if (namespaceOptional.isPresent()) {
+          getHistogram(resourceClass.get(), namespaceOptional.get(), name, execName, successType)
+              .update(toSeconds(startTime));
+          getCounter(
+                  resourceClass.get(),
+                  namespaceOptional.get(),
+                  name,
+                  execName,
+                  SUCCESS,
+                  successType)
+              .inc();
+        }
+      }
+      return result;
+    } catch (Exception e) {
```

Comment:
Is this a narrowest exception?

Reply:
This one was not narrowed down for metrics purpose only (so that it would be captured by the metrics). The exception is thrown at the end to ensure it's handled. There's a unit test added for this purpose as well.